### PR TITLE
Complete overhaul for the CTF script

### PR DESCRIPTION
**Pull request by Zanieon**
This PR is required for #829 to compile check properly, so imma already do it as well.

Yes yes @GeckoEidechse i know you hate me dumping all changes in one go, but i actually reordered some funcions and nested them with the ASCII art titles as i usually do, that means the diff will always be huge cuz github dumb to know code was reordered.

### CODE REVIEW

**Comment by GeckoEidechse** [comment]
Looked through this here and there over the past few days with the intention of splicing out individual changes and so far was rather unsuccessful in particular cause of how many lines this touches.

I'm really hesitant to merge any changes I don't fully understand so I'm trying to figure out the best way to merge this atm.
(I really want a Squirrel auto-formatter already so that I can at least weed out the formatting changes...)

Maybe we should go over the changes together in vc or something?

**Comment by Zanieon** [comment]
> Maybe we should go over the changes together in vc or something?

I'm down for it so i can tell what the changes does yeah.

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut [hunk 1]
```diff
@@ -334,29 +413,28 @@ void function CaptureFlag( entity player, entity flag )
...
 	MessageToTeam( GetOtherTeam( team ), eEventNotifications.PlayerCapturedFriendlyFlag, player, player )
 	EmitSoundOnEntityToTeam( flag, "UI_CTF_3P_EnemyScores", flag.GetTeam() )
 	
-	if ( GameRules_GetTeamScore( team ) == GameMode_GetRoundScoreLimit( GAMETYPE ) - 1 )
+	if ( GameRules_GetTeamScore( team ) == GetScoreLimit_FromPlaylist() - 1 )
 	{
 		PlayFactionDialogueToTeam( "ctf_notifyWin1more", team )
 		PlayFactionDialogueToTeam( "ctf_notifyLose1more", GetOtherTeam( team ) )
+		foreach( entity otherplayer in GetPlayerArray() )
+			Remote_CallFunction_NonReplay( otherplayer, "ServerCallback_CTF_PlayMatchNearEndMusic" )
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:429]:
```suggestion
		foreach( entity otherPlayer in GetPlayerArray() )
			Remote_CallFunction_NonReplay( otherPlayer, "ServerCallback_CTF_PlayMatchNearEndMusic" )
```

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut [hunk 2]
```diff
@@ -269,7 +347,8 @@ void function GiveFlag( entity player, entity flag )
 	flag.Signal( "ResetDropTimeout" )
 
 	flag.SetParent( player, "FLAG" )
-	thread DropFlagIfPhased( player, flag )
+	if( GetCurrentPlaylistVarInt( "phase_shift_drop_flag", 0 ) == 1 )
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:350]:
```suggestion
	if ( GetCurrentPlaylistVarInt( "phase_shift_drop_flag", 0 ) == 1 )
```

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut [hunk 3]
```diff
@@ -211,23 +246,62 @@ bool function VerifyCTFSpawnpoint( entity spawnpoint, int team )
...
+void function OnPlaying()
+{
+	foreach ( entity player in GetPlayerArray() )
+		CTFInitPlayer( player )
+}
+
 void function CTFInitPlayer( entity player )
 {
-	if ( !IsValid( file.imcFlagSpawn ) )
+	if( !GamePlaying() )
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:273]:
```suggestion
	if ( !GamePlaying() )
```

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut [hunk 4]
```diff
@@ -151,50 +160,76 @@ void function CreateFlags()
...
+			
+			rating += spawn.NearbyEnemyScore( team, "ai" )
+			rating += spawn.NearbyEnemyScore( team, "titan" )
+			rating += spawn.NearbyEnemyScore( team, "pilot" )
+		
+			rating = rating / allyFlagDistance
+		}
+
+		if ( spawn == player.p.lastSpawnPoint )
+			rating += GetConVarFloat( "spawnpoint_last_spawn_rating" )
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:225]: What's the default value for this convar? Should this be inverted to `-=` (We presumably don't want to spawn the player in the same place over and over, so if the default value is >0 it should be inverted)

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut [hunk 5]
```diff
@@ -1,5 +1,4 @@
 untyped
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:1]: Is `untyped` needed in this file? Which function calls here require it?

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Some nitpicks and some questions, logic overall looks good though

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut (hunk 4 above)
Comment (Zanieon) [reply Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:225]: Default value is `-4` so inverting would actually do what you just mentioned.

File: Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut (hunk 5 above)
Comment (Zanieon) [reply Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_ctf.nut:1]: i guess it's just because `flag.s.canTake` at this point.

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good!

**Review by GeckoEidechse — APPROVED** [review]
Confirmed working in testing when playing a few matches with @ASpoonPlaysGames @RoyalBlue1 @Alystrasz @eve-v0

**Comment by GeckoEidechse** [comment]
Aight, we spliced out as much as possible. For the last remaining changes, we decided in VC to merge it altogether and fix any remaining bugs post merge.
It was playtested in its current state so hopefully there's no major issues ^^

**Pull request merged**